src/main/scala/com/databricks/labs/overwatch/utils/PipelineValidationHelper.scala

```diff
@@ -0,0 +1,684 @@
+  private val logger: Logger = Logger.getLogger(this.getClass)
+
+  private var _pipelineSnapTime: Long = _
+
+  private val _healthCheck_id: String = java.util.UUID.randomUUID.toString
+  private val _quarantine_id: String = java.util.UUID.randomUUID.toString
+
+  private def healthCheckID: String = _healthCheck_id
+
+  def quarantineID: String = _quarantine_id
+
+  def etlDB: String = _etlDB
+
+  private var _validations: ArrayBuffer[HealthCheckReport] = new ArrayBuffer[HealthCheckReport]()
+  private var _quarantine: ArrayBuffer[QuarantineReport] = new ArrayBuffer[QuarantineReport]()
+
+  def validations: ArrayBuffer[HealthCheckReport] = _validations
+
+  def quarantine: ArrayBuffer[QuarantineReport] = _quarantine
+
+  val _isOverwatchDB: Boolean = spark.sessionState.catalog.getDatabaseMetadata(etlDB).properties.getOrElse("OVERWATCHDB", "FALSE").toBoolean
+
+  private def isOverwatchDB = _isOverwatchDB
+
+  private val workSpace = if (isOverwatchDB) {
+    println(s"$etlDB is Overwatch Database and Suitable for Pipeline Validation")
+    Helpers.getWorkspaceByDatabase(etlDB)
+  } else {
+    val errMsg = s"${etlDB} is Not Overwatch Database.Pipeline Validation can only work on Overwatch Database Tables. Validation aborted!!!"
+    throw new BadConfigException(errMsg)
+  }
+
+  val validPrefixes = List("dbfs:", "s3", "abfss", "gs")
+  val storagePrefix: String = workSpace.getConfig.etlDataPathPrefix
+  var healthCheckBasePath = storagePrefix.replace("global_share", "") + "healthCheck"
+  if (!validPrefixes.exists(healthCheckBasePath.startsWith)) {
+    healthCheckBasePath = s"dbfs:$healthCheckBasePath"
+  }
+
+  val healthCheckReportPath = s"""$healthCheckBasePath/heathCheck_report"""
+  val quarantineReportPath = s"""$healthCheckBasePath/quarantine_report"""
+
+  val Overwatch_RunIDs :Array[String] = if (spark.catalog.tableExists(s"$etlDB.pipeline_report")) {
+    val All_Overwatch_RunID = spark.read.table(s"$etlDB.pipeline_report")
+      .select("Overwatch_RunID").distinct().collect().map(_.getString(0))
+    if (Helpers.pathExists(healthCheckReportPath)) {
+      val healthCheckDF = spark.read.load(healthCheckReportPath)
+      val Validated_Overwatch_RunIDs = healthCheckDF.select("Overwatch_RunID").distinct().collect().map(_.getString(0))
+      All_Overwatch_RunID.diff(Validated_Overwatch_RunIDs)
+    } else {
+      All_Overwatch_RunID
+    }
+  }
+    else{
+    val errMsg = s"pipeline_report is not present in $etlDB. To proceed with pipeline_validation , pipeline_report table needs to be present in the database.Pipeline Validation aborted!!!"
+    throw new BadConfigException(errMsg)
+  }
+
+
+  private val gold = Gold(workSpace)
+  private val goldTargets = gold.GoldTargets
+
+
+  val jrcpKey: Array[String] = goldTargets.jobRunCostPotentialFactTarget._keys ++ goldTargets.jobRunCostPotentialFactTarget.partitionBy
+  val clsfKey: Array[String] = goldTargets.clusterStateFactTarget._keys ++ goldTargets.clusterStateFactTarget.partitionBy
```

Comment:
for line 82-90 please write a function which will return goldTargets.<target>._keys ++ goldTargets.<target>.partitionBy and use it everywhere

Reply:
@sriram251-code  Done.

  val jrcpKey: Array[String] = getKeysAndPartitionBy(goldTargets.jobRunCostPotentialFactTarget)
  val clsfKey: Array[String] = getKeysAndPartitionBy(goldTargets.clusterStateFactTarget)
  val jobRunKey: Array[String] = getKeysAndPartitionBy(goldTargets.jobRunTarget)
  val nbkey: Array[String] = getKeysAndPartitionBy(goldTargets.notebookTarget)
  val nbcmdkey: Array[String] = getKeysAndPartitionBy(goldTargets.notebookCommandsTarget)
  val clusterKey: Array[String] = getKeysAndPartitionBy(goldTargets.clusterTarget)
  val sparkJobKey: Array[String] = getKeysAndPartitionBy(goldTargets.sparkJobTarget)
  val sqlQueryHistKey: Array[String] = getKeysAndPartitionBy(goldTargets.sqlQueryHistoryTarget)
  val jobKey: Array[String] = getKeysAndPartitionBy(goldTargets.jobTarget)